File: kotlin/2359.kt
class Solution {
    fun closestMeetingNode(edges: IntArray, node1: Int, node2: Int): Int {
        val n = edges.size
        val dist1 = getDistances(edges, node1, n)
        val dist2 = getDistances(edges, node2, n)
        var minMaxDist = Int.MAX_VALUE
        var resultNode = -1
        for (i in 0 until n) {
            if (dist1[i] != -1 && dist2[i] != -1) {
                val currentMax = maxOf(dist1[i], dist2[i])
                if (currentMax < minMaxDist) {
                    minMaxDist = currentMax
                    resultNode = i
                }
            }
        }
        return resultNode
    }
    private fun getDistances(edges: IntArray, start: Int, n: Int): IntArray {
        val dist = IntArray(n) { -1 }
        var curr = start
        var d = 0
        while (curr != -1 && dist[curr] == -1) {
            dist[curr] = d++
            curr = edges[curr]
        }
        return dist
    }
}

File: tasks/2359.py
from typing import List

class Solution:
    def closestMeetingNode(self, edges: List[int], node1: int, node2: int) -> int:
        n = len(edges)

        dist1 = self.get_distances(edges, node1)
        dist2 = self.get_distances(edges, node2)

        answer = -1
        best_distance = 10 ** 9

        for i in range(n):
            if dist1[i] != -1 and dist2[i] != -1:
                current_distance = max(dist1[i], dist2[i])

                if current_distance < best_distance:
                    best_distance = current_distance
                    answer = i

        return answer

    def get_distances(self, edges: List[int], start: int) -> List[int]:
        n = len(edges)
        distances = [-1] * n

        current = start
        distance = 0

        while current != -1 and distances[current] == -1:
            distances[current] = distance
            distance += 1
            current = edges[current]

        return distances